src/charm.py

```diff
@@ -0,0 +1,293 @@
+
+        unit_data = event.relation.data[self.unit]
+        unit_data["certificate"] = cert
+        unit_data["fingerprint"] = self._get_fingerprint(cert)
+        unit_data["ip"] = listen_address
+        unit_data["port"] = str(self.config.get("port", ""))
+
+    def _get_fingerprint(self, cert: str) -> str:
+        x509_cert = x509.load_pem_x509_certificate(cert.encode())
+        return x509_cert.fingerprint(hashes.SHA256()).hex()
+
+    def _remove_all_certificates(self):
+        """Remove all old certificates held by AAR.
+
+        All client certificates are kept by juju in the relation until the hook
+        fails, so every time a new client is registered, the previous certificates
+        will be added again. To avoid duplicates and properly take into account
+        units that departed, we remove all certificates and add the ones that
+        are still active again.
+        """
+        shutil.rmtree(CLIENT_CERT_PATH)
+        shutil.rmtree(PUBLISHERS_CERT_PATH)
+        os.makedirs(CLIENT_CERT_PATH, 0o700, exist_ok=True)
+        os.makedirs(PUBLISHERS_CERT_PATH, 0o700, exist_ok=True)
+
+    def _on_aar_changed(self, event: ops.RelationChangedEvent):
+        self.unit.status = ops.MaintenanceStatus("Confguring new AMS Client")
+        self._remove_all_certificates()
+        ams_clients = []
+        for unit in event.relation.units:
+            cert = event.relation.data[unit].get("certificate")
+            if not cert:
+                continue
+
+            # If there is a relation mismatch, we set an error status
+            # ex: juju add-relation aar:client ams:publisher
+            mode = event.relation.data[unit]["mode"].strip('"')
+            if mode not in event.relation.name:
+                self.unit.status = ops.BlockedStatus(
+                    "Invalid relation {} to {}".format("arr-" + mode, event.relation.name)
+                )
+                return
+            ams_clients.append(event.relation.data[unit])
+
+        if not ams_clients:
+            return
+        for client in ams_clients:
+            self._register_ams_client(client["certificate"], event.relation.name)
+            logger.info(f"client registered {client['certificate']}")
+
+        aar_snap = snap.SnapCache()["aar"]
+        aar_snap.restart()
+        self.unit.status = ops.ActiveStatus()
+
+    def _register_ams_client(self, client_certificate: str, mode: str):
+        try:
+            client_certificate = client_certificate.replace("\\n", "\n").strip('"')
+            cmd = ["/snap/bin/aar", "trust", "add"]
+            if mode == "publisher":
+                cmd.append("--publisher")
+            check_output(cmd, stderr=STDOUT, input=client_certificate.encode("utf-8"))
+        except CalledProcessError as ex:
+            logger.error(f"failed to add client to aar: {ex.output}")
+            raise
+
+    def _snap_installed(self, snap_name) -> bool:
+        snap_client = snap.SnapClient()
```

Comment:
Does it make sense to create on `SnapClient` per instance of `AARCharm` and reuse?

Reply:
I was thinking to separate this out in a class of its own to manage the snap installation.